sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala

```diff
@@ -118,6 +122,95 @@ case class DataSourceResolution(
+          val staticPartitionProjectList = {
+            // check that the data column counts match
+            val numColumns = table.output.size
+            if (numColumns > staticPartitions.size + i.query.output.size) {
+              throw new AnalysisException(s"Cannot write: too many columns")
+            } else if (numColumns < staticPartitions.size + i.query.output.size) {
+              throw new AnalysisException(s"Cannot write: not enough columns")
+            }
+
+            val staticNames = staticPartitions.keySet
+
+            // for each static name, find the column name it will replace and check for unknowns.
+            val outputNameToStaticName = staticNames.map(staticName =>
+              table.output.find(col => resolver(col.name, staticName)) match {
+                case Some(attr) =>
+                  attr.name -> staticName
+                case _ =>
+                  throw new AnalysisException(
+                    s"Cannot add static value for unknown column: $staticName")
+              }).toMap
+
+            // for each output column, add the static value as a literal
+            // or use the next input column
+            val queryColumns = i.query.output.iterator
+            table.output.map { col =>
+              outputNameToStaticName.get(col.name).flatMap(staticPartitions.get) match {
+                case Some(staticValue) =>
+                  Alias(Cast(Literal(staticValue), col.dataType), col.name)()
+                case _ =>
+                  queryColumns.next
+              }
+            }
+          }
+
+          val dynamicPartitionOverwrite = table.table.partitioning.size > 0 &&
+            staticPartitions.size < table.table.partitioning.size &&
+            conf.partitionOverwriteMode == PartitionOverwriteMode.DYNAMIC
+
+          val query =
+            if (staticPartitions.isEmpty) {
+              i.query
+            } else {
+              Project(staticPartitionProjectList, i.query)
+            }
+
+          val deleteExpr =
+            if (staticPartitions.isEmpty) {
+              Literal(true)
+            } else {
+              staticPartitions.map { case (name, value) =>
+                query.output.find(col => resolver(col.name, name)) match {
+                  case Some(attr) =>
+                    EqualTo(attr, Cast(Literal(value), attr.dataType))
+                  case None =>
+                    throw new AnalysisException(s"Unknown static partition column: $name")
+                }
+              }.toSeq.reduce(And)
+            }
+
+          if (!i.overwrite) {
+            AppendData.byPosition(table, query)
+          } else if (dynamicPartitionOverwrite) {
+            OverwritePartitionsDynamic.byPosition(table, query)
+          } else {
+            OverwriteByExpression.byPosition(table, query, deleteExpr)
+          }
+        })
+        .getOrElse(i)
+
+    case i @ InsertTableStatement(UnresolvedRelation(AsTableIdentifier(_)), _, _, _, _)
```

Comment:
This also needs to handle the case where the identifier is owned by the session catalog (no default catalog, no catalog in the identifier), but the table provider is a v2 implementation. To do that, I think this should load the table using the v2 session catalog, then check the table instance. If it is a `CatalogTableAsV2`, then convert to the old `InsertIntoTable` like this rule currently does. Otherwise, it should convert to a v2 plan because it has a v2 table implementation.